{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "658532d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This notebook is used to check the normality distribution of \n",
    "# proteome and transcriptome data of the big data project\n",
    "# Inport all necessary packages\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import shapiro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f35b69fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check normality of the proteome data\n",
    "# First, read in the proteome data\n",
    "proteome_data = pd.read_csv(\"proteome_data_20201007.csv\")\n",
    "sample_name_list = ['S'+str(i) for i in range(1,28)]\n",
    "#proteome_data_normality_df=pd.DataFrame()\n",
    "#proteome_data_normality_df\n",
    "statistics=[]\n",
    "ps=[]\n",
    "normals=[]\n",
    "for name in sample_name_list:\n",
    "    data_without_na = proteome_data[name].dropna()\n",
    "    st,p_value=shapiro(data_without_na)\n",
    "    alpha = 0.05\n",
    "    if p_value>0.05:\n",
    "        normality=True\n",
    "    else:\n",
    "        normality=False\n",
    "    statistics.append(st)\n",
    "    ps.append(p_value)\n",
    "    normals.append(normality)\n",
    "proteome_data_normality_df = pd.DataFrame({'sample':sample_name_list,\n",
    "                                         'stats':statistics,\n",
    "                                         'p_val':ps,\n",
    "                                         'normality':normals\n",
    "                                          })\n",
    "proteome_data_normality_df.to_excel('proteome_data_normality_check.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "12a1c8e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\bioex\\Anaconda3\\lib\\site-packages\\scipy\\stats\\morestats.py:1681: UserWarning: p-value may not be accurate for N > 5000.\n",
      "  warnings.warn(\"p-value may not be accurate for N > 5000.\")\n"
     ]
    }
   ],
   "source": [
    "# Check normality of the transcripteome data\n",
    "# First, read in the transcripteome data\n",
    "transcripteome_data = pd.read_csv(\"Transcripteome20210906.csv\")\n",
    "sample_name_list = ['D025','D05','D1','D15','D2','D25','D3','D35','D4']\n",
    "\n",
    "statistics=[]\n",
    "ps=[]\n",
    "normals=[]\n",
    "for name in sample_name_list:\n",
    "    data_without_na = transcripteome_data[name].dropna()\n",
    "    st,p_value=shapiro(data_without_na)\n",
    "    alpha = 0.05\n",
    "    if p_value>0.05:\n",
    "        normality=True\n",
    "    else:\n",
    "        normality=False\n",
    "    statistics.append(st)\n",
    "    ps.append(p_value)\n",
    "    normals.append(normality)\n",
    "transcripteome_data_normality_df = pd.DataFrame({'sample':sample_name_list,\n",
    "                                         'stats':statistics,\n",
    "                                         'p_val':ps,\n",
    "                                         'normality':normals\n",
    "                                          })\n",
    "transcripteome_data_normality_df.to_excel('Transcripteome_data_normality_check.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "e0b90bcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check normality of individual mRNA across all dilution rates\n",
    "transcripteome_data_individual_mRNA = pd.read_excel(\"Supp_data3_transcriptome.xlsx\",index_col=0)\n",
    "transcripteome_data_individual_mRNA = transcripteome_data_individual_mRNA.T\n",
    "mRNA_name_list = transcripteome_data_individual_mRNA.columns.to_list()\n",
    "mRNA_name_list = mRNA_name_list[:-1]\n",
    "\n",
    "statistics=[]\n",
    "ps=[]\n",
    "normals=[]\n",
    "for name in mRNA_name_list:\n",
    "    data_without_na = transcripteome_data_individual_mRNA[name].dropna()\n",
    "    st,p_value=shapiro(data_without_na)\n",
    "    alpha = 0.05\n",
    "    if p_value>0.05:\n",
    "        normality=True\n",
    "    else:\n",
    "        normality=False\n",
    "    statistics.append(st)\n",
    "    ps.append(p_value)\n",
    "    normals.append(normality)\n",
    "individual_mRNA_data_normality_df = pd.DataFrame({'sample':mRNA_name_list,\n",
    "                                         'stats':statistics,\n",
    "                                         'p_val':ps,\n",
    "                                         'normality':normals\n",
    "                                          })\n",
    "individual_mRNA_data_normality_df.to_csv(\"Individual_mRNA_normality.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "7440d96b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check normality of individual protein across all dilution rates\n",
    "protein_data_individual_gene = pd.read_csv(\"proteome_data_20201007.csv\",index_col=3)\n",
    "protein_data_individual_gene = protein_data_individual_gene.T.iloc[4:-1,:]\n",
    "protein_name_list = protein_data_individual_gene.columns.to_list()\n",
    "\n",
    "valid_protein_names=[]\n",
    "statistics=[]\n",
    "ps=[]\n",
    "normals=[]\n",
    "for name in protein_name_list:\n",
    "    data_without_na = protein_data_individual_gene[name].dropna()\n",
    "    if len(data_without_na)<=3:\n",
    "        continue\n",
    "    st,p_value=shapiro(data_without_na)\n",
    "    alpha = 0.05\n",
    "    if p_value>0.05:\n",
    "        normality=True\n",
    "    else:\n",
    "        normality=False\n",
    "    valid_protein_names.append(name)\n",
    "    statistics.append(st)\n",
    "    ps.append(p_value)\n",
    "    normals.append(normality)\n",
    "individual_protein_data_normality_df = pd.DataFrame({'sample':valid_protein_names,\n",
    "                                         'stats':statistics,\n",
    "                                         'p_val':ps,\n",
    "                                         'normality':normals\n",
    "                                          })\n",
    "individual_protein_data_normality_df.to_csv(\"individual_protein_normality.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "709bd8e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "mRNA_data = pd.read_excel(\"pvsm.xlsx\",sheet_name='mRNA',index_col=0)\n",
    "protein_data = pd.read_excel(\"pvsm.xlsx\",sheet_name=\"Protome\",index_col=0)\n",
    "protein_data = protein_data.iloc[:,1:10]\n",
    "tmp_colnames = protein_data.columns\n",
    "protein_data.columns = ['D'+name[2:] if 'C_' in name else name for name in tmp_colnames]\n",
    "protein_data_T = protein_data.T\n",
    "mRNA_data_T = mRNA_data.T\n",
    "# Check normality of genes with high spearman correlation coefficient between protein and mRNA\n",
    "gene_name_list = ['YEL039C','YGR088W','YCR031C','YLR442C','YIR021W','YIL033C',\n",
    "                  'YNL302C','YLR438W','YFL036W','YDR256C','YLR393W','YPL119C',\n",
    "                  'YCL044C','YCR010C','YGR087C','YNL312W','YMR267W','YMR152W',\n",
    "                  'YGL169W','YGR214W','YKL135C','YKL026C','YKR049C','YKR076W',\n",
    "                  'YBR222C','YBR185C','YJR103W','YHR197W','YEL055C','YER145C',\n",
    "                  'YIL057C','YJL053W','YIL124W','YIR036C','YIR037W','YPL171C',\n",
    "                  'YDR195W','YJL054W','YJR074W','YJR134C','YOL033W','YML056C',\n",
    "                  'YMR175W','YGL236C','YGL228W','YGR248W','YNR040W','YNL274C',\n",
    "                  'YNL141W','YNL081C','YNL014W','YAL054C','YKR007W','YML131W',\n",
    "                  'YML127W','YMR278W','YHL007C','YMR298W','YMR090W','YMR107W',\n",
    "                  'YMR110C','YLR345W','YLR357W','YDR330W','YDL072C','YOR289W',\n",
    "                  'YDR056C','YLL023C','YLR093C','YDL204W','YOL124C']\n",
    "#check whether these gene's protein or mRNA distribution is normal\n",
    "valid_p_names=[]\n",
    "p_statistics=[]\n",
    "p_ps=[]\n",
    "p_normals=[]\n",
    "valid_m_names=[]\n",
    "m_statistics=[]\n",
    "m_ps=[]\n",
    "m_normals=[]\n",
    "\n",
    "for name in gene_name_list:\n",
    "    data_without_na = mRNA_data_T[name].dropna()\n",
    "    st,p_value=shapiro(data_without_na)\n",
    "    alpha = 0.01\n",
    "    if p_value>alpha:\n",
    "        normality=True\n",
    "    else:\n",
    "        normality=False\n",
    "    m_statistics.append(st)\n",
    "    m_ps.append(p_value)\n",
    "    m_normals.append(normality)\n",
    "    \n",
    "    data_without_na = protein_data_T[name].dropna()\n",
    "    if len(data_without_na)<=3:\n",
    "        continue\n",
    "    st,p_value=shapiro(data_without_na)\n",
    "    alpha = 0.01\n",
    "    if p_value>alpha:\n",
    "        normality=True\n",
    "    else:\n",
    "        normality=False\n",
    "    valid_p_names.append(name)\n",
    "    p_statistics.append(st)\n",
    "    p_ps.append(p_value)\n",
    "    p_normals.append(normality)   \n",
    "        \n",
    "highcorr_p_data_normality_df = pd.DataFrame({'sample':valid_p_names,\n",
    "                                         'stats':p_statistics,\n",
    "                                         'p_val':p_ps,\n",
    "                                         'normality':p_normals\n",
    "                                          })\n",
    "highcorr_m_data_normality_df = pd.DataFrame({'sample':gene_name_list,\n",
    "                                         'stats':m_statistics,\n",
    "                                         'p_val':m_ps,\n",
    "                                         'normality':m_normals\n",
    "                                          })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "a358f260",
   "metadata": {},
   "outputs": [],
   "source": [
    "highcorr_m_data_normality_df.to_csv('highcorr_mRNA_normality.csv')\n",
    "highcorr_p_data_normality_df.to_csv('highcorr_protein_normality.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
